Feature/timelines filter custom fields 3701

nkenneweg opened the pull request:
Implements [3701](https://www.openproject.org/work_packages/3701)

nkenneweg committed:
bug: featuresteps: custom field set,wp field check

nkenneweg committed:
timelines: custom field filters.

for now only for list and boolean custom fields.

nkenneweg committed:
intendation changed.

nkenneweg committed:
pass timelines_id for local custom fields filter.

nkenneweg committed:
integrative timelines custom field filter tests.

also adds these steps:
- the custom field <name> is enabled for the project <name>
- when i filter for work packages with the boolean custom field <name> set to <value>
- when i filter for work packages with the list custom field <name> set to <value>

nkenneweg committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_filter_custom_fields_3701

nkenneweg committed:
changelog.

nkenneweg committed:
use only custom fields for this project.

GarstgerUnhold commented:
Stuff's a-failin'. And looks like it can be related to this, so it's worth checking out before merging this.

Nilos commented:
I can not see the error-output on travis. Any Idea what is failing?

nkenneweg committed:
fix tests. user needs view work package permission.

apfelfabrik committed:
Merge remote-tracking branch 'origin/dev' into feature/timelines_filter_custom_fields_3701

Conflicts:
	doc/CHANGELOG.md

apfelfabrik commented:
Merging with dev branch, and testing on internal CI.

apfelfabrik commented on app/views/timelines/_custom_fields.html.erb:
```diff
@@ -0,0 +1,34 @@
+%>
+  <p>
+    <%= label_tag "timeline_options_custom_fields_#{custom_field.id}", custom_field.name %>
+    <% if custom_field.field_format == "bool" then %>
```
This field_format case select looks like it might as well be extracted into separate partials.

apfelfabrik commented:
CI is green, I hope travis is of the same opinion. I found a problem with filters for (none), though, which is documented in the ticket.

apfelfabrik committed:
Fixes filter for (none) in custom fields.

The id -1 and the empty string both need to be added as values in the
(none)-case, since (none) has to both match work packages with custom
fields with an empty value and work packages with no custom field.

apfelfabrik committed:
Fixes a bug w/ (none) combined with another selection.

Of course, (none) can be selected in a list of several other options, in
which case we need the empty string fix, too.
